File: sdk/javascript/src/services/urls.ts
import { HTTPClient } from '../http-client';
import { API_ENDPOINTS } from '../config';
import { 
  ShortURL, 
  CreateURLRequest, 
  UpdateURLRequest, 
  PaginationParams, 
  PaginatedResponse,
  RequestConfig 
} from '../types';

export class URLService {
  constructor(private client: HTTPClient) {}

  /**
   * Create a new short URL
   */
  async create(data: CreateURLRequest, config?: RequestConfig): Promise<ShortURL> {
    return this.client.post<ShortURL>(API_ENDPOINTS.URLS, data, config);
  }

  /**
   * Get all URLs for the authenticated user
   */
  async list(
    pagination?: PaginationParams,
    config?: RequestConfig
  ): Promise<PaginatedResponse<ShortURL>> {
    const params = new URLSearchParams();
    
    if (pagination?.limit) {
      params.set('limit', pagination.limit.toString());
    }
    if (pagination?.offset) {
      params.set('offset', pagination.offset.toString());
    }

    const url = `${API_ENDPOINTS.URLS}${params.toString() ? `?${params.toString()}` : ''}`;
    const response = await this.client.get<{
      urls: ShortURL[];
      total: number;
      limit: number;
      offset: number;
    }>(url, config);

    return {
      data: response.urls,
      total: response.total,
      limit: response.limit,
      offset: response.offset,
    };
  }

  /**
   * Get a specific URL by ID
   */
  async getById(id: number, config?: RequestConfig): Promise<ShortURL> {
    return this.client.get<ShortURL>(API_ENDPOINTS.URL_BY_ID(id), config);
  }

  /**
   * Update an existing URL
   */
  async update(id: number, data: UpdateURLRequest, config?: RequestConfig): Promise<ShortURL> {
    return this.client.put<ShortURL>(API_ENDPOINTS.URL_BY_ID(id), data, config);
  }

  /**
   * Delete a URL
   */
  async delete(id: number, config?: RequestConfig): Promise<void> {
    await this.client.delete(API_ENDPOINTS.URL_BY_ID(id), config);
  }

  /**
   * Get popular/trending URLs (public endpoint)
   */
  async getPopular(limit?: number, config?: RequestConfig): Promise<ShortURL[]> {
    const params = new URLSearchParams();
    if (limit) {
      params.set('limit', limit.toString());
    }

    const url = `${API_ENDPOINTS.URL_POPULAR}${params.toString() ? `?${params.toString()}` : ''}`;
    const response = await this.client.get<{ urls: ShortURL[] }>(url, config);
    return response.urls;
  }

  /**
   * Get URL by short code (for redirect purposes)
   */
  async getByShortCode(shortCode: string, config?: RequestConfig): Promise<ShortURL> {
    return this.client.get<ShortURL>(`/${shortCode}`, config);
  }

  /**
   * Validate a URL before shortening
   */
  async validateURL(url: string): Promise<boolean> {
    try {
      new URL(url);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Check if a custom alias is available
   */
  async checkAliasAvailability(alias: string, config?: RequestConfig): Promise<boolean> {
    try {
      await this.client.get(`/${alias}`, config);
      return false; // If we get a response, alias is taken
    } catch (error: any) {
      // If we get a 404, alias is available
      return error.statusCode === 404;
    }
  }

  /**
   * Bulk create URLs
   */
  async bulkCreate(urls: CreateURLRequest[], config?: RequestConfig): Promise<ShortURL[]> {
    return this.client.post<ShortURL[]>(`${API_ENDPOINTS.URLS}/bulk`, { urls }, config);
  }

  /**
   * Bulk delete URLs
   */
  async bulkDelete(ids: number[], config?: RequestConfig): Promise<void> {
    await this.client.post(`${API_ENDPOINTS.URLS}/bulk-delete`, { ids }, config);
  }

  /**
   * Search URLs by title or original URL
   */
  async search(
    query: string,
    pagination?: PaginationParams,
    config?: RequestConfig
  ): Promise<PaginatedResponse<ShortURL>> {
    const params = new URLSearchParams();
    params.set('q', query);
    
    if (pagination?.limit) {
      params.set('limit', pagination.limit.toString());
    }
    if (pagination?.offset) {
      params.set('offset', pagination.offset.toString());
    }

    const url = `${API_ENDPOINTS.URLS}/search?${params.toString()}`;
    const response = await this.client.get<{
      urls: ShortURL[];
      total: number;
      limit: number;
      offset: number;
    }>(url, config);

    return {
      data: response.urls,
      total: response.total,
      limit: response.limit,
      offset: response.offset,
    };
  }

  /**
   * Get URL statistics summary
   */
  async getStats(id: number, config?: RequestConfig): Promise<{
    totalClicks: number;
    uniqueClicks: number;
    clicksToday: number;
    clicksThisWeek: number;
    clicksThisMonth: number;
  }> {
    return this.client.get<{
      totalClicks: number;
      uniqueClicks: number;
      clicksToday: number;
      clicksThisWeek: number;
      clicksThisMonth: number;
    }>(`${API_ENDPOINTS.URL_BY_ID(id)}/stats`, config);
  }

  /**
   * Toggle URL active status
   */
  async toggleActive(id: number, config?: RequestConfig): Promise<ShortURL> {
    return this.client.post<ShortURL>(`${API_ENDPOINTS.URL_BY_ID(id)}/toggle`, {}, config);
  }

  /**
   * Duplicate an existing URL
   */
  async duplicate(id: number, config?: RequestConfig): Promise<ShortURL> {
    return this.client.post<ShortURL>(`${API_ENDPOINTS.URL_BY_ID(id)}/duplicate`, {}, config);
  }
}

File: sdk/python/urlshortener/types/urls.py
"""URL management related type definitions."""

from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field, HttpUrl, validator


class ShortURL(BaseModel):
    """Short URL model."""
    id: int = Field(description="URL ID")
    short_code: str = Field(description="Short URL code")
    original_url: HttpUrl = Field(description="Original long URL")
    title: Optional[str] = Field(None, description="URL title")
    description: Optional[str] = Field(None, description="URL description")
    custom_alias: Optional[str] = Field(None, description="Custom alias")
    password: Optional[str] = Field(None, description="Password protection")
    expires_at: Optional[datetime] = Field(None, description="Expiration date")
    is_active: bool = Field(default=True, description="Whether URL is active")
    click_count: int = Field(default=0, description="Total click count")
    user_id: int = Field(description="Owner user ID")
    created_at: datetime = Field(description="Creation timestamp")
    updated_at: datetime = Field(description="Last update timestamp")
    
    # Computed fields
    is_expired: bool = Field(default=False, description="Whether URL has expired")
    short_url: Optional[str] = Field(None, description="Full short URL")


class CreateURLRequest(BaseModel):
    """Request to create a new short URL."""
    original_url: HttpUrl = Field(description="Original long URL")
    title: Optional[str] = Field(None, max_length=255, description="URL title")
    description: Optional[str] = Field(None, max_length=1000, description="URL description")
    custom_alias: Optional[str] = Field(None, max_length=50, description="Custom alias")
    password: Optional[str] = Field(None, max_length=100, description="Password protection")
    expires_at: Optional[datetime] = Field(None, description="Expiration date")
    tags: Optional[List[str]] = Field(None, description="URL tags")
    
    @validator('custom_alias')
    def validate_custom_alias(cls, v):
        if v is not None:
            # Only allow alphanumeric, hyphens, and underscores
            import re
            if not re.match(r'^[a-zA-Z0-9_-]+$', v):
                raise ValueError('Custom alias can only contain letters, numbers, hyphens, and underscores')
        return v


class UpdateURLRequest(BaseModel):
    """Request to update an existing short URL."""
    title: Optional[str] = Field(None, max_length=255, description="URL title")
    description: Optional[str] = Field(None, max_length=1000, description="URL description")
    password: Optional[str] = Field(None, max_length=100, description="Password protection")
    expires_at: Optional[datetime] = Field(None, description="Expiration date")
    is_active: Optional[bool] = Field(None, description="Whether URL is active")
    tags: Optional[List[str]] = Field(None, description="URL tags")


class BulkCreateURLRequest(BaseModel):
    """Request to create multiple URLs."""
    urls: List[CreateURLRequest] = Field(description="List of URLs to create")


class BulkDeleteRequest(BaseModel):
    """Request to delete multiple URLs."""
    ids: List[int] = Field(description="List of URL IDs to delete")


class URLSearchRequest(BaseModel):
    """Request to search URLs."""
    query: str = Field(min_length=1, description="Search query")
    tags: Optional[List[str]] = Field(None, description="Filter by tags")
    is_active: Optional[bool] = Field(None, description="Filter by active status")
    created_after: Optional[datetime] = Field(None, description="Filter by creation date")
    created_before: Optional[datetime] = Field(None, description="Filter by creation date")


class URLStats(BaseModel):
    """URL statistics summary."""
    total_clicks: int = Field(description="Total click count")
    unique_clicks: int = Field(description="Unique visitor count")
    clicks_today: int = Field(description="Clicks today")
    clicks_this_week: int = Field(description="Clicks this week")
    clicks_this_month: int = Field(description="Clicks this month")
    last_clicked_at: Optional[datetime] = Field(None, description="Last click timestamp")
    
    # Geographic stats
    top_countries: List[str] = Field(default_factory=list, description="Top countries")
    top_cities: List[str] = Field(default_factory=list, description="Top cities")
    
    # Device stats
    top_devices: List[str] = Field(default_factory=list, description="Top devices")
    top_browsers: List[str] = Field(default_factory=list, description="Top browsers")
    
    # Referrer stats
    top_referrers: List[str] = Field(default_factory=list, description="Top referrers")


class URLPerformance(BaseModel):
    """URL performance metrics."""
    url: ShortURL = Field(description="Short URL details")
    clicks: int = Field(description="Total clicks")
    unique_clicks: int = Field(description="Unique clicks")
    click_rate: float = Field(description="Click rate percentage")
    last_clicked_at: Optional[datetime] = Field(None, description="Last click timestamp")
    performance_score: float = Field(description="Performance score (0-100)")


class PasswordValidationRequest(BaseModel):
    """Request to validate URL password."""
    password: str = Field(description="Password to validate")


class URLRedirectResponse(BaseModel):
    """URL redirect response."""
    original_url: HttpUrl = Field(description="Original URL to redirect to")
    redirect_type: str = Field(default="302", description="HTTP redirect type")
    requires_password: bool = Field(default=False, description="Whether password is required")


class AliasAvailabilityResponse(BaseModel):
    """Custom alias availability response."""
    alias: str = Field(description="Requested alias")
    available: bool = Field(description="Whether alias is available")
    suggestions: Optional[List[str]] = Field(None, description="Alternative suggestions")